include/joypad.h

```diff
@@ -0,0 +1,427 @@
+/**
+ * @file joypad.h
+ * @author Christopher Bonhage (me@christopherbonhage.com)
+ * @brief Joypad Subsystem
+ * @ingroup joypad
+ */
+
+#ifndef __JOYPAD_H
+#define __JOYPAD_H
+
+#include <assert.h>
+#include <stdbool.h>
+
+#include "joybus.h"
+
+/**
+ * @addtogroup joypad
+ * @{
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @brief Joypad Port Numbers */
+typedef enum
+{
+    /** @brief Joypad Port 1 */
+    JOYPAD_PORT_1     = 0,
+    /** @brief Joypad Port 2 */
+    JOYPAD_PORT_2     = 1,
+    /** @brief Joypad Port 3 */
+    JOYPAD_PORT_3     = 2,
+    /** @brief Joypad Port 4 */
+    JOYPAD_PORT_4     = 3,
+} joypad_port_t;
+
+/** @brief Count of Joypad ports */
+#define JOYPAD_PORT_COUNT 4
+
+/** @brief Convenience macro to validate a Joypad port number */
+#define ASSERT_JOYPAD_PORT_VALID(port) \
+    assert((port) >= 0 && (port) < JOYPAD_PORT_COUNT)
+
+/** @brief Convenience macro to iterate through all Joypad ports */
+#define JOYPAD_PORT_FOREACH(iterator_token) \
+    for (\
+        joypad_port_t iterator_token = JOYPAD_PORT_1; \
+        iterator_token < JOYPAD_PORT_COUNT; \
+        iterator_token += 1 \
+    )
```

Comment:
I would advocate for keeping this in the public API. It nicely encapsulates the logic for looping through all controller ports. Selfishly, I kept it public because I intend on using it in my projects, and it's used in the `joypadtest` example.

I am open to renaming it, but I would like to have a good reason for making it internal-only.

Reply:
Ok.